### PR TITLE
resholve: 0.9.1 -> 0.10.2, ongoing cross fixes

### LINKED ISSUE #292393: Build failure: xdg-utils (aarch64-linux cross)
### Steps To Reproduce

1. `nix-build -A pkgsCross.aarch64-multiplatform.xdg-utils -j1 --cores 16`

### Build log
A lot of similar failures via `binlore`/`yallback`/`resholve`.

```
building '/nix/store/7qdf9vjpizykfkwz7jfrdl9vvkp9d7sv-xdg-user-dirs-aarch64-unknown-linux-gnu-0.18-binlore.drv'...
generating binlore for /nix/store/7731h00maa974hx6hd1pa2cac14l3lii-xdg-user-dirs-aarch64-unknown-linux-gnu-0.18 by running:
/nix/store/g9cip59gy4vips6wfgnn8k0ds6ycw6ba-yara-aarch64-unknown-linux-gnu-4.4.0/bin/yara --scan-list --recursive /nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yar <(printf '%s\n' /nix/store/7731h00maa974hx6hd1pa2cac14l3lii-xdg-user-dirs-aarch64-unknown-linux-gnu-0.18/{bin,lib,libexec}) | /nix/store/6f8b71wahxqn95013rrbji8j7x44zl0n-yallback-aarch64-unknown-linux-gnu-0.2.0/bin/yallback /nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yall
/build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 21: /nix/store/g9cip59gy4vips6wfgnn8k0ds6ycw6ba-yara-aarch64-unknown-linux-gnu-4.4.0/bin/yara: cannot execute binary file: Exec format error
/nix/store/6f8b71wahxqn95013rrbji8j7x44zl0n-yallback-aarch64-unknown-linux-gnu-0.2.0/bin/.yallback-wrapped: line 4: shopt: progcomp: invalid shell option name
/nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yall: line 8: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/mktemp: cannot execute binary file: Exec format error
/nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yall: line 8: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/mktemp: cannot execute binary file: Exec format error
/nix/store/6f8b71wahxqn95013rrbji8j7x44zl0n-yallback-aarch64-unknown-linux-gnu-0.2.0/bin/.yallback-wrapped: line 31: compgen: command not found
/nix/store/6f8b71wahxqn95013rrbji8j7x44zl0n-yallback-aarch64-unknown-linux-gnu-0.2.0/bin/.yallback-wrapped: line 31: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/sort: cannot execute binary file: Exec format error
/nix/store/6f8b71wahxqn95013rrbji8j7x44zl0n-yallback-aarch64-unknown-linux-gnu-0.2.0/bin/.yallback-wrapped: line 159: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/sort: cannot execute binary file: Exec format error
/nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yall: line 54: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/sort: cannot execute binary file: Exec format error
/nix/store/pj1yyp67mw43x0wib284ywgifll6rink-source/execers.yall: line 55: /nix/store/f1j59l21z68vr1ds4garm518czs328nc-coreutils-aarch64-unknown-linux-gnu-9.4/bin/sort: cannot execute binary file: Exec format error
note: keeping build directory '/var/buildroot/nix-build-xdg-user-dirs-aarch64-unknown-linux-gnu-0.18-binlore.drv-3'
```

### Additional context
1. `xdg-utils` depends on procmail, which also fail to cross-build
2. Following change fix some issues, but uncover more `resholve` stuff
```diff
diff --git a/pkgs/development/tools/analysis/binlore/default.nix b/pkgs/development/tools/analysis/binlore/default.nix
index 54ea108b7d46..ecc4399a4641 100644
--- a/pkgs/development/tools/analysis/binlore/default.nix
+++ b/pkgs/development/tools/analysis/binlore/default.nix
@@ -1,8 +1,7 @@
 { lib
 , fetchFromGitHub
 , runCommand
-, yallback
-, yara
+, buildPackages
 }:
 
 /* TODO/CAUTION:
@@ -27,6 +26,7 @@ that down will almost certainly mean reworking the API.
 */
 
 let
+  inherit (buildPackages) yara yallback;
   src = fetchFromGitHub {
     owner = "abathur";
     repo = "binlore";
```
### Notify maintainers

/cc @K900
also /cc @emilylange because it affecting chromium cross-build (sorry for bothering you)

### Metadata
Verified on  575147aed746e3230b46795176e50645b8fe8387 (origin/master, origin/HEAD)

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### LINKED ISSUE #295072: lesspipe: cross-compilation regressed
### Steps To Reproduce

This is on nixos-unstable 3030f185:

```
> nix-build -A  pkgsCross.aarch64-multiplatform.lesspipe
...
Traceback (most recent call last):
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 4748, in <module>
    sys.exit(punshow())
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 947, in punshow
    resolve_cmdlikes()
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 3119, in resolve_cmdlikes
    cmdlike.resolve()
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 3091, in resolve
    self._resolve_invocations(solution)
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 3050, in _resolve_invocat
ions
    source.look_for_external_sub_execution(self.name, invocation)
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 4046, in look_for_externa
l_sub_execution
    if handler(parsed, invocation):
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 3717, in handle_external_
sed
    if self._find_sed_e_cmd(first):
  File "/nix/store/3ny4ai6x7warnq39hpspqjxxcfv2ibgs-resholve-0.9.1/bin/.resholve-wrapped", line 3659, in _find_sed_e_cmd
    close_fds=True,
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/subprocess.py", line 1046, in _execute_
child
    child_exception = pickle.loads(data)
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/pickle.py", line 1388, in loads
    return Unpickler(file).load()
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/pickle.py", line 977, in load_string
    self.append(rep.decode("string-escape"))
  File "/nix/store/mxwqwa270lp03z996zz6s6vy23cp67di-python-2.7.18.7/lib/python2.7/encodings/__init__.py", line 100, in se
arch_function
    level=0)
TypeError: _oildev_import() got an unexpected keyword argument 'fromlist'
/nix/store/4jq69bbd4fmb7gi48qbpx005ly6vgbvf-stdenv-linux/setup: line 131: pop_var_context: head of shell_variables not a
function context
error: builder for '/nix/store/khs61hf5q34vfbqlffdyn1lrnhcxlmxc-lesspipe-aarch64-unknown-linux-gnu-2.11.drv' failed with
exit code 1;
```

IDK why the log is truncated in the web view, it looks fine in markdown. idgi.

```
TypeError: _oildev_import() got an unexpected keyword argument 'fromlist'
```

### Build log

```
log here if short otherwise a link to a gist
```

### Additional context

cc: @abathur 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by abathur**
## Description of changes

Continuing from #297626 due to retarget pings (see approvals there; cc @danielbarter @K900 @thiagokokada).

AFAICT this will fix #295072 and fix #292393. I've tested cross for `lesspipe` and `xdg-utils` in CI (https://github.com/abathur/resholve/actions/runs/8474467054/job/23238072055).

I'm not certain about the approach here; if this is wrong I could use a little help understanding why. I've left review comments to annotate most of this to make it easier to see what's actually cross-related.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/misc/resholve/oildev.nix [hunk 1]
```diff
@@ -58,59 +38,54 @@ rec {
         hash on rev updates. Command will fail w/o and not print hash.
       */
       postFetch = ''
-        rm -rf $out/{Python-2.7.13,metrics,py-yajl,rfc,gold,web,testdata,services,demo,devtools}
+        rm -rf $out/{Python-2.7.13,metrics,py-yajl,rfc,gold,web,testdata,services,demo}
```
Comment (thiagokokada) [inline pkgs/development/misc/resholve/oildev.nix:41]: Small nit here, but could we extract the Python version from the package itself?

Alternatively, we could use `Python-*`.

**Review by thiagokokada — APPROVED** [review]

File: pkgs/development/misc/resholve/oildev.nix (hunk 1 above)
Comment (abathur) [reply pkgs/development/misc/resholve/oildev.nix:41]: This is just oil's vendored python

**Comment by abathur** [comment]
Going back to draft to evaluate the error reported in https://github.com/NixOS/nixpkgs/issues/292393#issuecomment-2021877222.

**Comment by abathur** [comment]
@danielbarter Can you give this another try?

**Comment by danielbarter** [comment]
yep, after rebasing this into staging, both `lesspipe` and `xdg-utils` build.

**Comment by abathur** [comment]
I left annotations on the previous PR (https://github.com/NixOS/nixpkgs/pull/297626#pullrequestreview-1956172613). Only material differences here should be the updated resholve rev and the lore tweak for xdg-utils.

**Review by danielbarter — APPROVED** [review]
Can't really comment on the resholve related details, but can confirm that both `xdg-utils` and `lesspipe` cross compiling (at least for aarch64) are fixed by this.

More generally, i think the resholve approach to improving the reliability of shell script packaging makes sense, but I am a little suspicious about the oil dependency. It is a fairly large project, with a vendored python 2 dependency (which is compiled to C++?!). If this is going to become the standard way for patching shell scripts in nixpkgs, it would be good to know that oil has a long term plan towards becoming, more boring?

**Comment by abathur** [comment]
> it would be good to know that oil has a long term plan towards becoming, more boring?

<details><summary>I guess this depends on your definition of boring :) (hiding since this is a tangent)</summary>

From resholve's prespective, I want a parser that is of sufficiently-high quality to drive a working shell, and ideally one that tries to do all it can statically at parse time. 

From this perspective, a very-complete parser that is good about failing at parse time when it's confused is quite boring compared to a much-less-complete parser that enables us to silently and subtly corrupt scripts as we rewrite them. 

I am not sure I would have arrived at them myself, but i think Andy's made a persuasive defense of his technical choices (and the ground he's covered also attests). 

As of the latest (0.21.0) release the c++ translation is finally passing all of the spec tests that the python impl passes. We still have to solve for interchange, whether thats just passing a parse tree representation or writing a python extension to drive the parser--but in either case it should enable resholve itself to get more boring by moving off of the python2 implementation.
</details>

**Review by thiagokokada — APPROVED** [review]

**Pull request merged**